diff -up cmake/PyQtMacros.cmake cmake/PyQtMacros.cmake.lib64
--- cmake/PyQtMacros.cmake	2019-08-06 15:58:04.255845979 +0200
+++ cmake/PyQtMacros.cmake.lib64	2019-08-06 15:58:52.428790521 +0200
@@ -36,7 +36,7 @@ MACRO(PYQT_WRAP_UI outfiles )
   ELSE(CMAKE_HOST_WIN32)
     # TODO osx
     SET(PYUIC_WRAPPER "${CMAKE_SOURCE_DIR}/scripts/pyuic-wrapper.sh")
-    SET(PYUIC_WRAPPER_PATH "${QGIS_OUTPUT_DIRECTORY}/lib")
+    SET(PYUIC_WRAPPER_PATH "${QGIS_OUTPUT_DIRECTORY}/lib${LIB_SUFFIX}")
   ENDIF(CMAKE_HOST_WIN32)
 
   FOREACH(it ${ARGN})
